Improvements

spkesan opened the pull request:
- This PR enables ECS task tags to be propagated as labels to the target info.
- This change includes addition of `--config.filter-ecs-tags` command line flag to filter specific ECS tags.
    - All tags are added by default.
    - `--config.filter-ecs-tags=""` disables all tags.
    - `--config.filter-ecs-tags="team,env"` allows just `team` and `env` tags

Commit:
Propagate ECS tags as labels

Commit:
Normalize label names for Prometheus

Commit:
Minor fix

Commit:
Allow filtering of ECS task tags

Commit:
Naming, comment tweak

Commit:
Fix config def

raags commented on main.go:
```diff
@@ -58,9 +44,11 @@ var outFile = flag.String("config.write-to", "ecs_file_sd.yml", "path of file to
 var interval = flag.Duration("config.scrape-interval", 60*time.Second, "interval at which to scrape the AWS API for ECS service discovery information")
 var times = flag.Int("config.scrape-times", 0, "how many times to scrape before exiting (0 = infinite)")
 var roleArn = flag.String("config.role-arn", "", "ARN of the role to assume when scraping the AWS API (optional)")
+var enableECSTags = flag.Bool("config.enable-ecs-tags", false, "Enable ECS tags to be propagated as labels during scrape.")
```
lets not have this flag, I doubt will not need discover ecs tags. Also if you really don't need to flag, you can use filterecstags and set a dummy filter that'll never match.

spkesan replied:
It appears clean with use of two configs. We should cover the following conditions,
- Enable tags, allow all tags by default
- Enable tags, filter tags using a filter list
- Disable tags completely

spkesan replied:
Removed `--config.enable-ecs-tags` config. Added separate validation to enable filter for ECS tags only if `--config.filter-ecs-tags` is used on CLI.
Now the behaviour is,
- By default allow all tags.
- `--config.filter-ecs-tags=""` disable all tags.
- `--config.filter-ecs-tags="team,env"` allow just `team` and `env` tags

raags commented on main.go:
```diff
@@ -58,9 +44,11 @@ var outFile = flag.String("config.write-to", "ecs_file_sd.yml", "path of file to
 var times = flag.Int("config.scrape-times", 0, "how many times to scrape before exiting (0 = infinite)")
 var roleArn = flag.String("config.role-arn", "", "ARN of the role to assume when scraping the AWS API (optional)")
+var enableECSTags = flag.Bool("config.enable-ecs-tags", false, "Enable ECS tags to be propagated as labels during scrape.")
+var filterECSTags = flag.String("config.filter-ecs-tags", "", "Specify ECS tags that will be allowed when \"config.enable-ecs-tags\" is enabled. Multiple tags should be comma separated, for example - \"team,aws:ecs:serviceName,aws:ecs:clusterName\". Default - allow all tags")
```
instead of `allow` just use the term `filter`. i.e. Specify CSV list of tags to filter and add as labels (defaults to all tags)

spkesan replied:
Replaced the term "allow" with "filter" across all the references for code consistency.

raags commented on main.go:
```diff
@@ -36,20 +36,6 @@ import (
-	DockerImage   string `yaml:"docker_image"`
-	MetricsPath   string `yaml:"__metrics_path__,omitempty"`
-	Scheme        string `yaml:"__scheme__,omitempty"`
-}
```
You can use the `inline` tag and not remove this struct : https://pkg.go.dev/gopkg.in/yaml.v2

spkesan replied:
This is exactly what I was looking for. Thanks. Made the changes.

raags replied:
why the formating change? Is this due to the use of tabs? Ensure that the editor treats tabs as spaces.

raags commented on main.go:
```diff
@@ -120,10 +108,22 @@ type PrometheusContainer struct {
 }
 
+// String replacer for normalizing label names
+var labelNameReplacer = strings.NewReplacer(".", "_", "-", "_", " ", "_", ":", "_")
```
Instead of specific chars, you can replace all invalid chars using a regex. Currently it may miss other chars for e.g. a tilde `~`. So replace all chars matching `[^A-Za-z_]`. The `^` is to negate the match.

Alos this var can be defined inside the function scope?

spkesan replied:
Perfect. Thanks, I further changed it to use the regex `[^a-zA-Z0-9_]` in accordance to Prometheus' label naming convention as it allows numbers too.

raags commented on main.go:
```diff
@@ -120,10 +108,22 @@ type PrometheusContainer struct {
+}
+
+// Allowed ECS Tags propagated as labels
+var allowedECSTags map[string]interface{}
```
shouldn't this be `filterECSTags` ?

spkesan replied:
This is a separate variable to store filtered ECS tags as a map. Anyway, I renamed the variable to `filteredECSTags`

raags reviewed (changes requested):
Also add description in the PR, you can also link the OPS ticket.

Commit:
Review changes

- Use inline field for tags instead of replacing the labels struct
- Use regex for normalizing label names
- Use term 'filter' over 'allow'

Commit:
Remove --config.enable-ecs-tags config. Instead enable filter only if --config.filter-ecs-tags is used

raags commented on main.go:
```diff
@@ -54,6 +69,21 @@ var prometheusServerNameLabel = flag.String("config.server-name-label", "PROMETH
+var filterECSTagsEnabled bool = false
+
+// isFlagPassed returns true if the flag has been used on the command line
+func isFlagPassed(name string) bool {
```
hm, weird that this is not part of the library, seems like would be a common use case.

raags commented on main.go:
```diff
@@ -124,6 +142,18 @@ type PrometheusTaskInfo struct {
 }
 
+// Regex used for normalizing label names
+var labelRE = regexp.MustCompile("[^a-zA-Z0-9_]")
```
why define this outside globally and not within the normalizeLabelName func?

spkesan replied:
I think it wouldn't make a significant difference. I can also define this inside `normalizeLabelName()`.

raags replied:
OK fine for now, but keeping variable inside functions they are used, helps in readability, and also limits the number of globals

Commit:
Document configurations (some missing configs too)

Commit:
Update Readme
